bindings/java/java/src/main/java/org/ethereum/evmc/EvmcVm.java

```diff
@@ -0,0 +1,135 @@
+package org.ethereum.evmc;
+
+import static org.ethereum.evmc.Host.addContext;
+import static org.ethereum.evmc.Host.removeContext;
+
+import java.nio.ByteBuffer;
+import java.util.Objects;
+
+/**
+ * The Java interface to the evm instance.
+ *
+ * <p>Defines the Java methods capable of accessing the evm implementation.
+ */
+public final class EvmcVm implements AutoCloseable {
+
+  private static EvmcVm evmcVm;
+  private static boolean isEvmcLibraryLoaded = false;
+  private ByteBuffer nativeVm;
+  /**
+   * This method loads the specified evm shared library and loads/initializes the jni bindings.
+   *
+   * @param filename /path/filename of the evm shared object
+   */
+  public static EvmcVm create(String filename) {
+    if (!EvmcVm.isEvmcLibraryLoaded) {
+      try {
+        // load so containing the jni bindings to evmc
+        System.load(System.getProperty("user.dir") + "/../c/build/evmc.so");
+        EvmcVm.isEvmcLibraryLoaded = true;
+      } catch (UnsatisfiedLinkError e) {
+        System.err.println("Native code library failed to load.\n" + e);
+        System.exit(1);
+      }
+    }
+    if (Objects.isNull(evmcVm)) {
+      evmcVm = new EvmcVm(filename);
+    }
+    return evmcVm;
+  }
+
+  private EvmcVm(String filename) {
+    // initialize jni and load EVM shared library
+    nativeVm = init(filename);
+  }
+
+  /**
+   * This method loads the specified evm implementation and initializes jni
+   *
+   * @param filename path + filename of the evm shared object to load
+   * @return
+   */
+  public native ByteBuffer init(String filename);
+
+  /**
+   * EVMC ABI version implemented by the VM instance.
+   *
+   * <p>Can be used to detect ABI incompatibilities. The EVMC ABI version represented by this file
+   * is in ::EVMC_ABI_VERSION.
+   */
+  public native int abi_version();
+
+  /**
+   * The name of the EVMC VM implementation.
+   *
+   * <p>It MUST be a NULL-terminated not empty string. The content MUST be UTF-8 encoded (this
+   * implies ASCII encoding is also allowed).
+   */
+  public native String name(ByteBuffer nativeVm);
```

Comment:
After the change this does not look right. Why can't be the `nativeVm` taken from the `this` object for every method implementation?

Besides, I believe the `jlong` will be enough to store the pointer to the EVMC VM.

Reply:
>  After the change this does not look right. Why can't be the nativeVm taken from the this object for every method implementation?

I'll make a wrapper method for each of the native methods that uses `this`

> Besides, I believe the jlong will be enough to store the pointer to the EVMC VM.

ByteBuffer imo is the better way to store pointers.  It is servers as a native pointer + some bookkeeping info.

Reply:
done